<a href="https://colab.research.google.com/github/avihai1002/Avihai-_Git_-Ex1/blob/main/PythonModule.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Env Setup

- Configure your python interpreter: https://www.jetbrains.com/help/pycharm/creating-virtual-environment.html 
- Python file vs Python console and notebooks 



This module inspired by https://www.udemy.com/course/complete-python-bootcamp/ 


# Debuggning your code in PyCharm
https://www.jetbrains.com/help/pycharm/debugging-your-first-python-application.html

# Further Reading

#### Same id for python different immutable python objects
https://stackoverflow.com/questions/21091568/why-two-individually-created-immutable-objects-have-same-id-and-mutable-objects

#### Python `super()` method
https://stackoverflow.com/questions/576169/understanding-python-super-with-init-methods


#### Python static varialbe and methods 
https://stackoverflow.com/questions/68645/static-class-variables-and-methods-in-python


#### Python design patterns
https://python-patterns.guide/



#### Python regular expressions 
https://realpython.com/regex-python/



# Python Katas

**Kata** is a Japanese word (型 or 形) meaning "form". It refers to a detailed choreographed pattern of martial arts movements made to be practised *alone*.

During this module, you are obliged to earn **at least 55 Katas**

- New python questions will be published once a week (Friday)
- Each question is 1 Kata, some are 2
- Each one of you has an opportunity to write Python unittests for the weekly questions set, and **earn 10 Katas** for that. Unittests must be submitted through a PR in Github and reviewed by a course stuff (Alon or Jenya). On PR completion, you should merge branch `main` to your branch called `python_katas_<alias>` (change <alias\> to your name) in which you will solve all your Katas
- Katas are published regardless the big exercises that are published from time to time
- In order to earn a Kata, you must pass the unittests. Your solutions are randomly tested manually by a course stuff, if you fool a test, you will be loss 10 Katas


# Testers weekly lottery

In [ ]:
students = [
    'AdanBhsas', 'AlexeyMihaylovDev', 'Amitpoz', 'Daniel-Reuven', 'danielbar0101', 'danielmalka14',
    'Danishain', 'Ddady1', 'dmitriyshub', 'dorondollev', 'DustyMadDude', 'edenganon',
    'eliranshriki', 'elkan316', 'Gershoz', 'Haimr101', 'itayb12', 'JohnSchiff',
    'kostalubarsky', 'mkshamir', 'netanalm', 'poratnick', 'RamiKandov', 'shayalon', 'Shaykc',
    'shaysoso', 'shlomigd', 'xXxARLxXx', 'yakirfrid', 'yosefdudi', 'zoharn007'
]

In [ ]:
import random
random.shuffle(students)

print('Testers for Kata 1:\n')
students[:7]



Testers for Kata 1:



['AdanBhsas',
 'RamiKandov',
 'netanalm',
 'yosefdudi',
 'elkan316',
 'DustyMadDude',
 'danielmalka14']

# Sematic Version 

Please read versioning semantic: 
https://semver.org/

Explore Flask Changelog: https://flask.palletsprojects.com/en/2.1.x/changes/  
Find an example for patch, minor and major changes.





# `pylint`

`pylint` tests for style as well as some very basic program logic.

First, if you don't have it already (and you probably do, as it's part of the Anaconda distribution), you should install `pylint`.<br>Once that's done feel free to comment out the cell, you won't need it anymore.

In [ ]:
! pip install pylint

Let's save a very simple script:

In [ ]:
%%writefile simple1.py
a = 1
b = 2
print(a)
print(B)

Overwriting simple1.py


Now let's check it using pylint

In [ ]:
! pylint simple1.py

************* Module simple1
C:  4, 0: Final newline missing (missing-final-newline)
C:  1, 0: Missing module docstring (missing-docstring)
C:  1, 0: Invalid constant name "a" (invalid-name)
C:  2, 0: Invalid constant name "b" (invalid-name)
E:  4, 6: Undefined variable 'B' (undefined-variable)

---------------------------------------------------------------------

Your code has been rated at -12.50/10 (previous run: 8.33/10, -20.83)





No config file found, using default configuration


Pylint first lists some styling issues - it would like to see an extra newline at the end, modules and function definitions should have descriptive docstrings, and single characters are a poor choice for variable names.

More importantly, however, pylint identified an error in the program - a variable called before assignment. This needs fixing.

Note that pylint scored our program a negative 12.5 out of 10. Let's try to improve that!

In [ ]:
%%writefile simple1.py
"""
A very simple script.
"""

def myfunc():
    """
    An extremely simple function.
    """
    first = 1
    second = 2
    print(first)
    print(second)

myfunc()

Overwriting simple1.py


In [ ]:
! pylint simple1.py

************* Module simple1
C: 14, 0: Final newline missing (missing-final-newline)

---------------------------------------------------------------------

Your code has been rated at 8.33/10 (previous run: -12.50/10, +20.83)





No config file found, using default configuration


Much better! Our score climbed to 8.33 out of 10. Unfortunately, the final newline has to do with how jupyter writes to a file, and there's not much we can do about that here. Still, pylint helped us troubleshoot some of our problems. But what if the problem was more complex?

In [ ]:
%%writefile simple2.py
"""
A very simple script.
"""

def myfunc():
    """
    An extremely simple function.
    """
    first = 1
    second = 2
    print(first)
    print('second')

myfunc()

Overwriting simple2.py


In [ ]:
! pylint simple2.py

************* Module simple2
C: 14, 0: Final newline missing (missing-final-newline)
W: 10, 4: Unused variable 'second' (unused-variable)

------------------------------------------------------------------

Your code has been rated at 6.67/10 (previous run: 6.67/10, +0.00)





No config file found, using default configuration


pylint tells us there's an unused variable in line 10, but it doesn't know that we might get an unexpected output from line 12! For this we need a more robust set of tools. That's where `unittest` comes in.